File: java-gdb-activities-20/activity19/src/com/gdb/logging/TransactionLogger.java
package com.gdb.logging;

import com.gdb.command.TransactionCommand;
import java.util.Collections;
import java.util.List;

/**
 * Bridge abstraction class decoupling high-level logger API from pluggable storage backends.
 */
public class TransactionLogger {
    protected LogDestination destination;

    public TransactionLogger(LogDestination destination) {
        this.destination = destination;
    }

    public void setDestination(LogDestination destination) {
        this.destination = destination;
    }

    public void log(TransactionCommand cmd) {
        if (destination != null) {
            destination.write(cmd);
        }
    }

    public List<TransactionCommand> readAll() {
        if (destination != null) {
            return destination.readAll();
        }
        return Collections.emptyList();
    }

    public void clear() {
        if (destination != null) {
            destination.clear();
        }
    }

    public String getDestinationName() {
        return destination != null ? destination.getDestinationName() : "UNKNOWN";
    }
}

File: python-gdb-activities-20/activity20-solution/gdb/logging/transaction_logger.py
# gdb/logging/transaction_logger.py
from typing import List
from gdb.domain.transaction import Transaction
from gdb.logging.log_destination import LogDestination

class TransactionLogger:
    """Bridge Abstraction: delegates logging operations to LogDestination implementor."""

    def __init__(self, destination: LogDestination) -> None:
        self._destination = destination

    def log(self, transaction: Transaction) -> None:
        if transaction is not None:
            self._destination.write(transaction)

    def get_transactions(self) -> List[Transaction]:
        return self._destination.read_all()

    def getTransactions(self) -> List[Transaction]:
        return self.get_transactions()

    def get_destination(self) -> LogDestination:
        return self._destination

    def getDestination(self) -> LogDestination:
        return self.get_destination()

    def set_destination(self, destination: LogDestination) -> None:
        self._destination = destination

    def setDestination(self, destination: LogDestination) -> None:
        self.set_destination(destination)
